impls/src/tor/config.rs

```diff
@@ -0,0 +1,341 @@
+
+	let ret = BASE32.encode(&address_bytes);
+	Ok(ret.to_lowercase())
+}
+
+pub fn create_onion_service_sec_key_file(
+	os_directory: &str,
+	sec_key: &DalekSecretKey,
+) -> Result<(), Error> {
+	let key_file_path = &format!("{}{}{}", os_directory, MAIN_SEPARATOR, SEC_KEY_FILE);
+	let mut file = File::create(key_file_path).context(ErrorKind::IO)?;
+	// Tag is always 32 bytes, so pad with null zeroes
+	file.write("== ed25519v1-secret: type0 ==\0\0\0".as_bytes())
+		.context(ErrorKind::IO)?;
+	let expanded_skey: ExpandedSecretKey = ExpandedSecretKey::from(sec_key);
+	file.write_all(&expanded_skey.to_bytes())
+		.context(ErrorKind::IO)?;
+	Ok(())
+}
+
+pub fn create_onion_service_pub_key_file(
+	os_directory: &str,
+	pub_key: &DalekPublicKey,
+) -> Result<(), Error> {
+	let key_file_path = &format!("{}{}{}", os_directory, MAIN_SEPARATOR, PUB_KEY_FILE);
+	let mut file = File::create(key_file_path).context(ErrorKind::IO)?;
+	// Tag is always 32 bytes, so pad with null zeroes
+	file.write("== ed25519v1-public: type0 ==\0\0\0".as_bytes())
+		.context(ErrorKind::IO)?;
+	file.write_all(pub_key.as_bytes()).context(ErrorKind::IO)?;
+	Ok(())
+}
+
+pub fn create_onion_service_hostname_file(os_directory: &str, hostname: &str) -> Result<(), Error> {
+	let file_path = &format!("{}{}{}", os_directory, MAIN_SEPARATOR, HOSTNAME_FILE);
+	let mut file = File::create(file_path).context(ErrorKind::IO)?;
+	file.write_all(&format!("{}.onion\n", hostname).as_bytes())
+		.context(ErrorKind::IO)?;
+	Ok(())
+}
+
+pub fn create_onion_auth_clients_dir(os_directory: &str) -> Result<(), Error> {
+	let auth_dir_path = &format!("{}{}{}", os_directory, MAIN_SEPARATOR, AUTH_CLIENTS_DIR);
+	fs::create_dir_all(auth_dir_path).context(ErrorKind::IO)?;
+	Ok(())
+}
+/// output an onion service config for the secret key, and return the address
+pub fn output_onion_service_config(
+	tor_config_directory: &str,
+	sec_key: &SecretKey,
+) -> Result<String, Error> {
+	let (_, d_pub_key) = ed25519_keypair(&sec_key)?;
+	let address = onion_address(&d_pub_key)?;
+	let hs_dir_file_path = format!(
+		"{}{}{}{}{}",
+		tor_config_directory, MAIN_SEPARATOR, HIDDEN_SERVICES_DIR, MAIN_SEPARATOR, address
+	);
+
+	// If file already exists, don't overwrite it, just return address
+	if Path::new(&hs_dir_file_path).exists() {
+		return Ok(address);
+	}
+
+	// create directory if it doesn't exist
+	fs::create_dir_all(&hs_dir_file_path).context(ErrorKind::IO)?;
+
+	let (d_sec_key, d_pub_key) = ed25519_keypair(&sec_key)?;
+	create_onion_service_sec_key_file(&hs_dir_file_path, &d_sec_key)?;
+	create_onion_service_pub_key_file(&hs_dir_file_path, &d_pub_key)?;
+	create_onion_service_hostname_file(&hs_dir_file_path, &address)?;
+	create_onion_auth_clients_dir(&hs_dir_file_path)?;
+
+	//TODO: Windows
+	fs::set_permissions(&hs_dir_file_path, fs::Permissions::from_mode(0o700))
+		.context(ErrorKind::IO)?;
+
+	Ok(address)
+}
+
+/// output torrc file given a list of hidden service directories
+pub fn output_torrc(
```

Comment:
This approach works fine, but it won't work as well for on-the-fly changes, like the user requesting a different receive address. In Grin++, I'm opening a torrc socket and using the AUTHENTICATE, ADD_ONION, etc APIs to allow that sort of thing to work more seamlessly. Maybe that's not necessary for grin-wallet, especially since we haven't switched to the wallet713-style logged-in CLI, but just wanted to point out that potential limitation.

Reply:
Discussed in keybase, this is mostly because there are no decent client libs in Rust, and this is much easier to implement for the time being (2 advantages in that it doesn't open any extra ports on the listener side and doesn't require keeping track of a password for the control port). Will keep as is for now and see how we get on.